pepper-apis/src/main/java/org/broadinstitute/ddp/customexport/db/dao/CustomExportDao.java

```diff
@@ -0,0 +1,84 @@
+        DBUtils.checkUpdate(1, lastCompletedByStudyId);
+    }
+
+    @SqlUpdate("insert into custom_export (study_id, custom_export_last_completed) values (:studyId, :lastCompleted)")
+    void insertExport(
+            @Bind("studyId") long studyId,
+            @Bind("lastCompleted") Instant lastCompleted
+    );
+
+    @SqlQuery("select * from custom_export where study_id = :id")
+    @RegisterConstructorMapper(CustomExportDto.class)
+    Optional<CustomExportDto> findExportDtoByStudyId(@Bind("id") long id);
+
+    @SqlQuery("select custom_export_last_completed from custom_export where study_id = :id")
+    Optional<Instant> findExportLastCompletedByStudyId(@Bind("id") long id);
+
+    @SqlUpdate("update custom_export set custom_export_last_completed = :lastCompleted where study_id = :studyId")
+    int updateLastCompletedByStudyId(@Bind("lastCompleted") Instant lastCompleted, @Bind("studyId") long studyId);
+
+    @SqlQuery("SELECT CASE WHEN ( "
+            + "SELECT COUNT(activity_instance_id) FROM ( "
+            + "SELECT aci.activity_instance_id FROM activity_instance AS aci "
+            + "JOIN study_activity AS sa ON sa.study_activity_id = aci.study_activity_id  "
+            + "JOIN activity_instance_status AS ais ON aci.activity_instance_id = ais.activity_instance_id  "
+            + "JOIN activity_instance_status_type AS aist ON aist.activity_instance_status_type_id = ais.activity_instance_status_type_id "
+            + "WHERE sa.study_id = :studyId AND aist.activity_instance_status_type_code=:statusType "
+            + "AND aci.first_completed_at>:lastCompletion AND sa.study_activity_code=:activityCode) as ids "
+            + ") > "
+            + "0 THEN true ELSE false END")
+    boolean needCustomExport(@Bind("studyId") long studyId,
+                             @Bind("statusType") String statusType,
+                             @Bind("lastCompletion") long lastCompletion,
+                             @Bind("activityCode") String activityCode);
+
+    @SqlQuery("SELECT usen.user_id as user_id, aci.first_completed_at as completed_time FROM user_study_enrollment AS usen "
+            + "JOIN study_activity AS sa ON sa.study_id = usen.study_id "
+            + "JOIN activity_instance AS aci ON usen.user_id = aci.participant_id "
+            + "AND aci.study_activity_id = sa.study_activity_id "
+            + "JOIN activity_instance_status AS ais ON aci.activity_instance_id = ais.activity_instance_id "
+            + "JOIN activity_instance_status_type AS aist ON aist.activity_instance_status_type_id = ais.activity_instance_status_type_id "
+            + "WHERE usen.study_id = :studyId AND aist.activity_instance_status_type_code=:statusType "
+            + "AND aci.first_completed_at>:lastCompletion AND usen.valid_to IS NULL AND sa.study_activity_code=:activityCode "
+            + "GROUP BY usen.user_id, aci.first_completed_at ORDER BY MAX(aci.first_completed_at) DESC LIMIT :limit OFFSET :offset")
```

Comment:
Have we considered sorting in `ASC` order instead? I think it will be safer that way. If we're in the middle of paging through the result set, and a couple more participants finish their survey, won't that throw off our iteration since we're effectively starting at the back of the iteration? (I think this is called "phantom" paging or something, not sure...) If we sort by ASC, then any newly finished ones will be at the end of the iteration and the results we have paged through so far won't be affected.

Reply:
Good point!